include/property.h

```diff
@@ -0,0 +1,181 @@
+#ifndef __PROPERTY_H__
+#define __PROPERTY_H__
+
+#include <iostream>
+#include <set>
+#include <tuple>
+#include <sstream>
+#include <regex>
+
+namespace clayer {
+
+    // Concepts used for property reading and writing
+    template<typename T>
+    concept bool Streamable = requires(T o, std::ostream& s) {
+        {s << o} -> std::ostream&;
+    };
+
+    template<typename T>
+    concept bool StringRef = requires(T o) {
+        {*o} -> std::string;
+    };
+
+    // template<typename T=void>
+    // void print<>(){
+    //     std::cout << "\n";
+    // }
+
+    // template<Streamable T, typename... Targs>
+    // void print(T o, Targs... tail) {
+    //     std::cout << o <<" ";
+    //     print(tail...);
+    // }
+
+    // ## Log LogRecord - Any item that can be printed in a Log Record
+    enum log_properties {
+        FILE,   // - File name (CodeContext)
+        FUNC,   // - Function name (CodeContext)
+        LINE,   // - Line number (CodeContext)
+        // LEVEL
+        // NAME
+        // THREAD
+        // HASH
+        DATE,   // - Date a Log Record is created (RunContext)
+        TIME,   // - Time a Log Record is created (RunContext)
+        MESG    // - Added by the user
+    };
+
+    // ### Code Context - Log Properties that can be inferred from Code
+    struct CodeContext {
+        std::string file;
+        std::string func;
+        int line;
+
+        CodeContext() = default;
+        CodeContext(const std::string& a, const std::string& b, const int& c): file(a), func(b), line(c) {}
+
+        decltype(auto) tie() const{
+            return std::tie(file, func, line);
+        }
+        bool operator<(const CodeContext& rhs) const {
+            return tie() < rhs.tie();
+        }
+    };
+    static const std::string CodeContextStart ("");
+    static const std::string CodeContextSep (" ");
+    static const std::string CodeContextEnd ("");
+    std::ostream& operator<<(std::ostream& s, const CodeContext& c) {
+        return s << CodeContextStart
+            << c.file << CodeContextSep
+            << c.func << CodeContextSep
+            << c.line << CodeContextEnd;
+    }
+
+    // TODO: RunContext
+
+    class LogRecord {
+    public:
+        using State = CodeContext;
+        LogRecord() : code() {};
+
+        friend std::ostream& operator<<(std::ostream& s, const LogRecord& c);
+
+        CodeContext code;
+        std::string message;
+
+        State get_state() {
+            return code;
+        }
+
+    private:
+        // TODO: make code private and try to make read_prop friendly
+        // NOTE: lmgtfy.com?q=one cannot declare a namespace within a class 
```

Comment:
* read_prop uses private methods
* when I tried to add read_prop as method of the class, compiler said you can't do the overloading trick if the function the function is not in a name space -- weird!!
* this means that read_prop should be inside a name space, when I tried to add namespace inside a class, looks like you can't declare a namespace within a class.